osh/hub/scan/mock.py

```diff
@@ -0,0 +1,133 @@
+    """
+    returns all arches available for given build
+    """
+    arches = []
+    for child_id in koji_proxy.getTaskDescendents(task_id, request=True):
+        child = koji_proxy.getTaskInfo(child_id, request=True)
+        if child['method'] == 'buildArch' and child['arch'] != 'noarch':
+            arches.append(child['arch'])
+
+    # noarch tasks can be built anywhere
+    return arches or koji_proxy.getAllArches()
+
+
+def _get_tag_specific_config(target):
+    """
+    returns build tag specific mock configuration which should be appended to
+    the generated mock config
+    """
+    # Fedora
+    match = re.search(r'f(\d+)', target)
+    if match is not None:
+        return f"config_opts['releasever'] = '{match[1]}'\n" \
+               f"config_opts['bootstrap_image'] = 'registry.fedoraproject.org/fedora:{match[1]}'\n"
+
+    # CentOS Stream
+    match = re.search(r'c(\d+)s', target)
+    if match is not None:
+        return f"config_opts['releasever'] = '{match[1]}'\n" \
+               f"config_opts['bootstrap_image'] = 'quay.io/centos/centos:stream{match[1]}'\n"
+
+    # RHEL
+    match = re.search(r'rhel-?(\d+)', target)
+    if match is not None:
+        return f"config_opts['releasever'] = '{match[1]}'\n" \
+               f"config_opts['bootstrap_image'] = 'registry.access.redhat.com/ubi{match[1]}/ubi'\n"
+
+    # fallback
+    return "config_opts['use_bootstrap_image'] = False\n"
+
+
+def generate_mock_configs(nvr, koji_profile, task_dir):
+    # retrieve the build
+    cfg = koji.read_config(koji_profile)
+    koji_proxy = koji.ClientSession(cfg['server'])
+    build = koji_proxy.getBuild(nvr)
+
+    # retrieve task parameters
+    task = koji_proxy.getTaskInfo(build['task_id'], request=True)
+    method = task['method']
+    params = koji.parse_task_params(method, task['request'])
+
+    # parse build tag name from task parameters
+    if method == 'build':
+        tag = _get_build_method_build_tag(koji_proxy, params['target'])
+    elif method == 'wrapperRPM':
+        tag = params['build_target']['build_tag_name']
+    else:
+        # FIXME: add actual error handling
+        logger.error(f'No build target for "{nvr}" available!')
+        return
+
+    # generate a config for every built arch
+    logger.debug(f'Generating mock configs for build tag "{tag}"')
+    for arch in _get_build_arches(koji_proxy, task['id']):
+        p = subprocess.run(['koji', '-p', koji_profile, 'mock-config',
+                            '--arch', arch, '--latest', '--tag', tag],
+                           check=True, stdout=subprocess.PIPE)
+        contents = p.stdout.decode()
+
+        # add extra repos
+        matched_repos = [repo for regex, repo
+                         in settings.MOCK_AUTO_EXTRA_REPOS.items()
+                         if re.search(regex, tag)]
+        extra_repos = '\n'.join(matched_repos).replace('\n', '\\n')
+        contents = contents.replace("[main]", extra_repos + '\\n\\n[main]')
```

Comment:
Could we please make `config_opts['root']` unique as well?  The locking mechanism in `csmock` assumes that each mock config uses its own directory for the (ch)root.  If we run two distinct mock configs that use the same value for `config_opts['root']`, the locking mechanism does not work:
```
[...]
>>> 2024-02-19 14:29:12	"/usr/bin/mock" "-r" "/tmp/tmpyfvruenl/mock.cfg" "--plugin-option=tmpfs:keep_mounted=True" "--config-opts=print_main_output=True" "--init"
INFO: mock.py version 5.2 starting (python version = 3.9.18, NVR = mock-5.2-1.kdudka.5.el9), args: /usr/libexec/mock/mock -r /tmp/tmpyfvruenl/mock.cfg --plugin-option=tmpfs:keep_mounted=True --config-opts=print_main_output=True --init
Start(bootstrap): init plugins
INFO: selinux enabled
Finish(bootstrap): init plugins
Start: init plugins
INFO: selinux enabled
Finish: init plugins
INFO: Signal handler active
Start: run
Start: clean chroot
ERROR: Build root is locked by another process.

!!! 2024-02-19 14:29:13	warning: failed to init mock profile (/tmp/tmpyfvruenl/mock.cfg), trying to scrub cache...
[...]
```

Rather than fixing the locking mechanism in `csmock` and making the tasks block each other, we can use separate (ch)roots and make the tasks finally run in parallel.

Reply:
In that case, we must implement some clean-up logic, because some chroot remains will be left on the filesystem when the task finishes.

Reply:
Good point.  `csmock` should optionally invoke `mock ... --scrub=all` before releasing the lock.  Something like https://github.com/openscanhub/openscanhub/blob/77f43ad4d0c0860ce1b72f9edae6717847a66003/osh/worker/csmock_runner.py#L123

In any case, this problem will go away once OSH starts to create a fresh worker for each task (similarly to how COPR works).

Reply:
Random root directory names are handled in ea34cbc3cef206f26c4d5fad92232cbc86a02e98.

Reply:
Both `csmock` pull requests are now merged.  I'll add the support for `--scrub-all` ASAP.

Reply:
Addressed in 74e85ca87ac515862ac2996fec87ee353acbddb4.

Reply:
The jobs will fail because there has not been a release of `csmock` yet.